test/integration/garbagecollector/garbage_collector_test.go

```diff
@@ -216,106 +176,49 @@ func setup(t *testing.T, workerCount int) *testContext {
+	ctx := util.ExtensionSetup(t, result, workerCount)
 	alwaysStarted := make(chan struct{})
 	close(alwaysStarted)
 	gc, err := garbagecollector.NewGarbageCollector(
-		clientSet,
-		metadataClient,
-		restMapper,
+		ctx.ClientSet,
+		ctx.MetadataClient,
+		ctx.RESTMapper,
 		garbagecollector.DefaultIgnoredResources(),
-		informerfactory.NewInformerFactory(sharedInformers, metadataInformers),
+		informerfactory.NewInformerFactory(ctx.SharedInformers, ctx.MetadataInformers),
 		alwaysStarted,
 	)
 	if err != nil {
 		t.Fatalf("failed to create garbage collector: %v", err)
 	}
 
-	stopCh := make(chan struct{})
-	tearDown := func() {
-		close(stopCh)
-		result.TearDownFn()
-	}
 	syncPeriod := 5 * time.Second
 	startGC := func(workers int) {
 		go wait.Until(func() {
 			// Resetting the REST mapper will also invalidate the underlying discovery
 			// client. This is a leaky abstraction and assumes behavior about the REST
 			// mapper, but we'll deal with it for now.
-			restMapper.Reset()
-		}, syncPeriod, stopCh)
-		go gc.Run(workers, stopCh)
-		go gc.Sync(clientSet.Discovery(), syncPeriod, stopCh)
+			ctx.RESTMapper.Reset()
+		}, syncPeriod, ctx.StopCh)
+		go gc.Run(workers, ctx.StopCh)
+		go gc.Sync(ctx.ClientSet.Discovery(), syncPeriod, ctx.StopCh)
 	}
 
 	if workerCount > 0 {
 		startGC(workerCount)
 	}
 
 	return &testContext{
-		tearDown:           tearDown,
+		tearDown:           ctx.TearDown,
 		gc:                 gc,
-		clientSet:          clientSet,
-		apiExtensionClient: apiExtensionClient,
-		dynamicClient:      dynamicClient,
-		metadataClient:     metadataClient,
+		clientSet:          ctx.ClientSet,
+		apiExtensionClient: ctx.APIExtensionClient,
+		dynamicClient:      ctx.DynamicClient,
+		metadataClient:     ctx.MetadataClient,
 		startGC:            startGC,
 		syncPeriod:         syncPeriod,
 	}
 }
 
-func createNamespaceOrDie(name string, c clientset.Interface, t *testing.T) *v1.Namespace {
-	ns := &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: name}}
-	if _, err := c.CoreV1().Namespaces().Create(context.TODO(), ns, metav1.CreateOptions{}); err != nil {
-		t.Fatalf("failed to create namespace: %v", err)
-	}
-	falseVar := false
-	_, err := c.CoreV1().ServiceAccounts(ns.Name).Create(context.TODO(), &v1.ServiceAccount{
-		ObjectMeta:                   metav1.ObjectMeta{Name: "default"},
-		AutomountServiceAccountToken: &falseVar,
-	}, metav1.CreateOptions{})
-	if err != nil {
-		t.Fatalf("failed to create service account: %v", err)
-	}
-	return ns
-}
-
-func deleteNamespaceOrDie(name string, c clientset.Interface, t *testing.T) {
-	zero := int64(0)
-	background := metav1.DeletePropagationBackground
-	err := c.CoreV1().Namespaces().Delete(context.TODO(), name, metav1.DeleteOptions{GracePeriodSeconds: &zero, PropagationPolicy: &background})
-	if err != nil {
-		t.Fatalf("failed to delete namespace %q: %v", name, err)
-	}
-}
-
 func TestCrossNamespaceReferencesWithWatchCache(t *testing.T) {
```

Comment:
could you add a test case or two to test?
1. CRD is installed, things working as expected 
2. then uninstalled, things working as expected
3. then reinstalled, things working as expected

Reply:
Take a look at the existing `TestCRDDeletionCascading`. If I'm understanding you correctly, that is exactly what this test is already doing. It has been the primary test I've used to confirm my GC changes do/don't work and it:
1. Installs a CRD
2. Creates a CR with a dependent on a builtin type
3. Uninstalls the CRD
4. Confirms that both the CR and dependent are deleted
5. Reinstalls the CRD and repeats steps 2-4.

Reply:
I've added a check to assert that the CRD informer must stop in a reasonable amount of time after CRD uninstall. 

https://github.com/kubernetes/kubernetes/pull/97214/files#diff-869f9a4b37327530c4d7890e790937adc33b0033dbad0a1ea0739f121f971d31R1175-R1182